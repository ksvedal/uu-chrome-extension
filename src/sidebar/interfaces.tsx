import { ReactNode } from "react";

export interface RadioButtonGroupProps {
  presetOption: string;
  index: number;
  onOptionChange: (option: string, index: number) => void;
}

export interface ToggleButtonProps {
  isChecked: boolean;
  onToggle: () => void;
  text: string;
}

export interface RegularButtonProps {
  text: string;
  onClick: () => void;
}

export interface CollapsibleArrowProps {
  isExpanded: boolean;
}

export interface CollapsibleItemTypeInterface {
  type: ElementType;
  thisElement: ElementObject | null;
  children?: ReactNode;
  isAllHighlighted: boolean;
  setIsAllHighlighted: React.Dispatch<React.SetStateAction<boolean>>;
  setCurrentHighlighted: React.Dispatch<React.SetStateAction<ElementObject | null>>;
  parentIndex: number;
  url: string;
  testID: string;
}

export interface CollapsibleItemElementInterface {
  type: ElementType;
  thisElement: ElementObject;
  children?: ReactNode;
  highlightedElement: ElementObject | null;
  isAllHighlighted: boolean;
  setHighlightedElement: React.Dispatch<React.SetStateAction<ElementObject | null>>;
  setIsAllHighlighted: React.Dispatch<React.SetStateAction<boolean>>;
  /* updateJson: (elementObject: ElementObject, index: number, url: string) => void;
  index: number;
  url: string;
  testregelId: string; */
}

export interface ResultsHeaderInterface {
  url: string;
  isScanned: boolean;
}

/**
 * Interface for an elementType
 * Example: Button, Header, Image, Link, etc.
 */
export interface ElementType {
  name: string;
  nodes: ElementObject[];
  selector: string;
}

/**
 * Interface for an elementObject
 * Type is an elementType
 * Title can be the text on the element
 * TODO: test out different cases of title, What should we do with images? Long links? etc. Most likely an alt_text or aria label can be used here
 *          Or just type + index
 * html is the html code for the element as a string
 */
export interface ElementObject {
  title: string;
  element: string;
  selector: string;
  result: ElementResult;
  attributes: ElementAttribute[];
  isCommentVisible: boolean;
}

export interface ElementResult {
  testregelId: string;
  nettlesar: string | null;
  utvidelse: string | null;
  side: string;
  element: string;
  samsvar: string;
  utfall: string;
  kommentar: string;
}
export interface ElementAttribute {
  name: string;
  value: string;
}

export interface ExtendedElementObject {
  isCommentVisible: boolean;
}