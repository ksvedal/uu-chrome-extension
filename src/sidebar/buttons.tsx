import React, { useState } from "react";
import { ToggleButtonProps, RegularButtonProps, CollapsibleArrowProps, RadioButtonGroupProps} from "./interfaces";

export const ToggleButton: React.FC<ToggleButtonProps> = ({ isChecked, onToggle, text }) => {
  return (
    <button className={`toggle-button float-right ${isChecked ? 'pressed' : ''}`} onClick={(e) => { e.stopPropagation(); onToggle(); }}>
      {text}
    </button>
  );
};

export const RegularButton: React.FC<RegularButtonProps> = ({ onClick, text }) => {
  return (
    <button className='regular-button float-right' onClick={(e) => { e.stopPropagation(); onClick(); }}>
      {text}
    </button>
  );
};

export const RadioButtonGroup: React.FC<RadioButtonGroupProps> = ({ onOptionChange }) => {
  const [selectedOption, setSelectedOption] = useState<string>('');

  const handleOptionChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const option = event.target.value;
    setSelectedOption(option);
    onOptionChange(option);
  };

  return (
    <div>
      <p>Does the button's text identify its function?</p>
      <label className={`radio-button ${selectedOption === 'Yes' ? 'active' : ''}`}>
        <input
          type="radio"
          name="option"
          value="Yes"
          checked={selectedOption === 'Yes'}
          onChange={handleOptionChange}
        />
        <span className="radio-button-text">Yes</span>
      </label>
      <label className={`radio-button ${selectedOption === 'No' ? 'active' : ''}`}>
        <input
          type="radio"
          name="option"
          value="No"
          checked={selectedOption === 'No'}
          onChange={handleOptionChange}
        />
        <span className="radio-button-text">No</span>
      </label>
      <label className={`radio-button ${selectedOption === 'The element is not a button' ? 'active' : ''}`}>
        <input
          type="radio"
          name="The element is not a button"
          value="The element is not a button"
          checked={selectedOption === 'The element is not a button'}
          onChange={handleOptionChange}
        />
        <span className="radio-button-text">The element is not a button</span>
      </label>
    </div>
  );
};





export const CollapsibleArrowButton: React.FC<CollapsibleArrowProps> = ({ isExpanded }) => {
  return <span className={`arrow-button ${isExpanded ? 'pressed' : ''}`}>{isExpanded ? ' ▲' : ' ▲'}</span>;
};
