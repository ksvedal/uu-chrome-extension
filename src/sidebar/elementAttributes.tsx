import React, { useEffect, useState } from 'react';
import { ElementAttribute, ElementObject } from './interfaces';

export const ElementAttributes: React.FC<ElementObject> = ({ attributes, htmlString, }) => {
    // Define your desired attribute names
    const attributeNames = ["aria-labelledby", "aria-label", "title", "Description", "Role", "Focusable"];

    const initialAttributes: ElementAttribute[] = attributeNames.map(name => ({ name, value: '' }));

    const [attributess, setAttributes] = useState<ElementAttribute[]>(initialAttributes);

    useEffect(() => {
        const newAttributes = attributess.map(attr => {
            const match = attributes.find(item => item.name === attr.name);
            return match ? match : attr;
        });
        setAttributes(newAttributes);
    }, [attributes]);

    //Table created by ElementObject to display attributes
    return (
        <div>
            <table>
                <thead>
                <tr>
                    <th className="tableHead attribute">Attribute</th>
                    <th className="tableHead value">Value</th>
                </tr>
                </thead>
                <tbody>
                {
                    // map over the attributess
                    attributess.map((item, index) =>
                        <AttributeField key={index} name={item.name} value={item.value} />
                    )
                }
                </tbody>
            </table>

        </div>
    );
};

export const AttributeField: React.FC<ElementAttribute> = ({ name, value }) => {
    //Table created by ElementObject to display attributes
    return (
        <tr>
            <td className="tableBody attribute">
                {name}
            </td>
            <td className="tableBody value">
                {value}
            </td>
        </tr>
    );
};
