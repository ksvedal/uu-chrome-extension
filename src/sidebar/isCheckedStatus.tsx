import React from 'react';

interface Props {
  text: string;
}

const IsCheckedStatus: React.FC<Props> = ({ text }) => {
  return (
    <div className={`${(text === "Yes") ? 'yesColor' : (text === "No") ? "noColor": (text === "Not a button") ? "notButtonColor": ""}`}>
        {text}
    </div>
  );
};

export default IsCheckedStatus;