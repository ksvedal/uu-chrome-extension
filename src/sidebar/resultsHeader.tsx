import React, { useContext } from 'react';
import TabIcon from './tabIcon';
import { ResultsHeaderInterface } from './interfaces';
import { MyContext } from './resultItemsContext';
import { TestUtils } from './testUtils';

const ResultsHeader: React.FC<ResultsHeaderInterface> = ({url, isScanned}) => {
  if(!isScanned){
    return null;
  }
const context = useContext(MyContext);
    if (context === null) {
      // handle the case where the context is null
      return null;
    }
    const { elementResults, setElementResults } = context;

    
    
    const logResult= () => {
      for (let result of elementResults){
        TestUtils.giveIdChromeAndExtensionVersion(result);
        console.log(result);
      };
      console.log("Result: " + elementResults);
    }
  
  return (
    <div className='row headline'>
        <div className={"col-2"}>
          <TabIcon></TabIcon>
        </div>
        <div className={"col-7"}>
            <p> {url} </p>
        </div>
        <div className={"col-3"}>
            <button className='headline-button float-right' onClick={logResult}>
                Print results
            </button>
        </div>
    </div>
  );
};

export default ResultsHeader;