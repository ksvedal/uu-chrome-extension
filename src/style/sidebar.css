/* --------------------------------------------------------------
| Design 3
-------------------------------------------------------------- */
@import url('https://fonts.googleapis.com/css?family=Open+Sans');

/* Variables */
:root {
  --primary-color: #f5f5f5;
  --secondary-color: #ffffff;
  --tertiary-color: #565656;

  --purple: #94227D;
  --light-purple: #9a618f;
  --orange: #ffa046;
  --light-orange: #ffc691;

  --text-color: #ffffff;
  --text-color-inverse: #000000;

  --good-color: #4ea289;
  --alert-color: #fddc6e;
  --danger-color: #ff2a5c;

  --very-small-font: 0.8rem;
  --small-font: 1rem;
  --medium-font: 2rem;
  --big-font: 3rem;

  --border: 0px solid var(--light-purple);
  --radius: 4px;

  --small-space: 0.5rem;
  --medium-space: 1rem;
  --big-space: 4rem;

  --small-padding: 0.3rem;
  --medium-padding: 0.8rem;
  --big-padding: 1rem;
}

body {
  margin: 0;
  padding: 0;
  background-color: var(--primary-color);
  color: var(--text-color-inverse);
  font-family: "Open Sans", sans-serif;
  font-weight: bold;
}

* {
  box-sizing: border-box;
}

/* Small: */
[class*="col-"] {
  width: 100%;
  float: left;
  padding: var(--small-padding);
}

/* Full width: */
@media only screen and (min-width: 400px) {
  .col-1 {
    width: 8.33%;
  }

  .col-2 {
    width: 16.66%;
  }

  .col-3 {
    width: 25%;
  }

  .col-4 {
    width: 33.33%;
  }

  .col-5 {
    width: 41.66%;
  }

  .col-6 {
    width: 60%;
  }

  .col-7 {
    width: 58.33%;
  }

  .col-8 {
    width: 66.66%;
  }

  .col-9 {
    width: 75%;
  }

  .col-10 {
    width: 83.33%;
  }

  .col-11 {
    width: 91.66%;
  }

  .col-12 {
    width: 100%;
  }
}

.row {
  margin-top: var(--small-space);
  padding: var(--small-padding);
}

.row::after {
  content: "";
  clear: both;
  display: table;
}

.float-left {
  float: left;
}

.float-right {
  float: right;
}

.whitebox {
  background-color: var(--secondary-color);
  border-radius: var(--radius);
  overflow-wrap: break-word;
  box-shadow: 0 1px 2px rgba(0, 0, 0, 0.15);
}

/* Header Field
---------------------------------------------- */
.header-field {
  background-color: var(--secondary-color);
  color: #0f69ef;
  display: flex;
}

.header-field .extension-text p {
  font-size: 2rem;
  margin-bottom: var(--medium-space);
}

.header-field .extension-logo {
  margin-left: 20px;
  margin-right: 10px;
}

.header-field .extension-logo img {
  display: block;
  width: 60px;
  margin-block-start: 1em;
  margin-block-end: 1em;
  margin-inline-start: 0px;
  margin-inline-end: 0px;
}

/* Scan-page field
---------------------------------------------- */
.scan-page-field {
  justify-content: center;
}

.scan-page-field button {
  padding: var(--medium-padding);
  border-radius: var(--radius);
  border: var(--border);
  background: var(--purple);
  color: var(--text-color);
  box-shadow: 0 3px 4px rgba(0, 0, 0, 0.12);
  font-family: "Open Sans", sans-serif;
  font-weight: bold;
  font-size: var(--very-small-font);
}

.scan-page-field button:hover {
  cursor: pointer;
}

.scan-page-field button:active {
  background: var(--primary-color);
}

/* Headline 
---------------------------------------------- */
.headline .headline-button {
  padding: var(--medium-padding);
  width: auto;
  border-radius: var(--radius);
  border: var(--border);
  background: var(--purple);
  color: var(--text-color);
  box-shadow: 0 3px 4px rgba(0, 0, 0, 0.12);
  font-family: "Open Sans", sans-serif;
  font-weight: bold;
  font-size: var(--very-small-font);
}

.headline .headline-button:hover {
  cursor: pointer;
}

.headline .headline-button:active {
  background: var(--primary-color);
}

.headline img {
  margin-top: var(--small-padding);
  height: 3rem;
  width: 3rem;
}

/* Collapsible content 
---------------------------------------------- */

/* Parent collapsible */
.collapsible-item {
}

.collapsible-item-parent {
  margin-top: var(--small-space);
  border-radius: var(--radius);
  background-color: white;
  padding: var(--small-padding);
  box-shadow: 0 1px 2px rgba(0, 0, 0, 0.15);
  transition: box-shadow 0.02s ease-in-out;
}

.collapsible-item-parent .item-header {
  padding: var(--medium-padding);
  align-items: center;
}

.collapsible-item-parent .arrow-button {
  font-size: 20px;
  color: var(--text-color-inverse);
  rotate: 90deg;
  padding: var(--small-padding);
}

.collapsible-item-parent .toggle-button {
  padding: var(--medium-padding);
  width: auto;
  border-radius: var(--radius);
  box-shadow: 0 3px 4px rgba(0, 0, 0, 0.12);
  background: var(--purple);
  color: var(--text-color);
  font-size: var(--very-small-font);
}

.total-buttons {
  align-content: center;
  align-items: center;
}

.buttons-text {
  font-size: var(--small-font);
  align-content: center;
  align-items: center;
}

.checkbox-button.pressed {
}

.checkbox-button.error {
  background: var(--tertiary-color);
  border: var(--border);
  box-shadow: 0 3px 4px rgba(0, 0, 0, 0.12);
}

.checkbox-button.error.pressed {
  background-color: var(--danger-color);
  color: var(--text-color-inverse);
  box-shadow: inset 0 3px 4px rgba(0, 0, 0, 0.12);
}

.checkbox-button.checked {
  background: var(--tertiary-color);
  border: var(--border);
  box-shadow: 0 3px 4px rgba(0, 0, 0, 0.12);
}

.checkbox-button.checked.pressed {
  background-color: var(--alert-color);
  color: var(--text-color-inverse);
  box-shadow: inset 0 3px 4px rgba(0, 0, 0, 0.12);
}

/* Parent collapsible on hover */
.collapsible-item-parent:hover {
  cursor: pointer;
  box-shadow: 0 3px 4px rgba(0, 0, 0, 0.12);
}

.collapsible-item-parent .col-3 .item-header:hover .arrow-button {
  transition: transform 0.1s ease;
  color: #94227d;
}

.collapsible-item-parent .item-header .toggle-button:hover {
  cursor: pointer;
}

/* Parent collapsible on click */
.collapsible-item .item-header.pressed {
}

.collapsible-item-parent .item-header.pressed .arrow-button {
  transition: transform 0.3s ease;
  transform: rotate(90deg);
  color: #94227d;
}

.collapsible-item-parent .item-header.pressed .total-buttons {
}

.collapsible-item-parent .item-header .toggle-button.pressed {
  background: var(--orange);
  box-shadow: inset 0 3px 4px rgba(0, 0, 0, 0.12);
}

/* Child collapsible */
.collapsible-item-children {
}

.collapsible-item-children .item-header {
  align-items: center;
  padding: var(--small-padding);
  border-radius: var(--radius);
  box-shadow: 0 3px 4px rgba(0, 0, 0, 0.12);
  background-color: var(--primary-color);
  margin-left: var(--small-space);
  margin-right: var(--small-space);
}

.collapsible-item-children .item-header .flex-item {
}

.collapsible-item-children .item-header .arrow-button {
  font-size: 17px;
  color: #aaaaaa;
  rotate: 90deg;
}

.collapsible-item-children .item-header .toggle-button {
  padding: var(--medium-padding);
  border-radius: var(--radius);
  border: var(--border);
  background: var(--purple);
  color: var(--text-color);
  font-family: "Open Sans", sans-serif;
  font-weight: bold;
  font-size: var(--very-small-font);
}

.collapsible-item-child .item-header .toggle-button:hover {
  cursor: pointer;
  background: var(--light-purple);
}

/* Child collapsible on hover */
.collapsible-item-child .item-header:hover {
  cursor: pointer;
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.12);
}

.collapsible-item-child .item-header:hover .arrow-button {
  transition: transform 0.3s ease;
  color: #94227d;
}

/* Child collapsible on click */
.collapsible-item-child .item-header.pressed {
  box-shadow: inset 0 3px 4px rgba(0, 0, 0, 0.12);
  background-color: var(--light-orange);
}

.collapsible-item-child .item-header.pressed .arrow-button {
  transition: transform 0.1s ease;
  transform: rotate(90deg);
  color: #94227d;
}

.collapsible-item-child .item-header.pressed .toggle-button {
}

.collapsible-item-child .item-header .toggle-button.pressed {
  background: var(--orange);
}


/* Content data */
.collapsible-item-child .content-data {
}

.tableBody {
  background-color: #d8bfd3;
  padding: var(--small-padding);
  border-radius: var(--radius);
  table-layout: fixed; /* Added */
  cursor: text;
}

.tableHead {
  height: var(--medium-space);
  padding: var(--small-padding);
  background-color: #a85e99;
  color: white;
  border-collapse: collapse;
  border-spacing: 0;
  border: var(--border);
  border-radius: var(--radius);
  table-layout: fixed;
  cursor: text;
}

.tableHead.attribute {
  min-width: 100px;
  width: 1%;
}

.tableHead.value {
  width: auto; /* Let the value column fill the remaining width */
}

.tss {
  background-color: #ff2a5c;
}

/* Comment box */
.collapsible-item-child .comment-box {
  border-radius: var(--radius);
  box-shadow: inset 0 3px 4px rgba(0, 0, 0, 0.12);
  width: 100%;

}
.collapsible-item-child .comment-box .textarea {
  border-radius: var(--radius);
  border: var(--border);
  box-shadow: inset 0 3px 4px rgba(0, 0, 0, 0.12);
  width: 100%;
  max-width: 100%;
  min-width: 100%;
  min-height: 100px;
  font-family: "Open Sans", sans-serif;
  font-weight: bold;
  padding: var(--medium-padding);
  border: 2px solid var(--light-orange);
}