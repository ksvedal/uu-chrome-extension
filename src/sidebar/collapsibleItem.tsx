import React, { useContext, useEffect, useState } from "react";
import { CollapsibleItemElementInterface, CollapsibleItemTypeInterface, ElementObject, ElementResult } from "./interfaces";
import { ToggleButton, CollapsibleArrowButton } from "./buttons";
import { MessageSender } from "../messageObjects/messageSender";
import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
import { vs } from 'react-syntax-highlighter/dist/esm/styles/prism';
import { ElementAttributes } from "./elementAttributes";
import { MyContext } from "./resultItemsContext";

const messageSender = new MessageSender();

export const CollapsibleItemType: React.FC<CollapsibleItemTypeInterface> = ({ type, thisElement, index }) => {
    const [currentHighlighted, setCurrentHighlighted] = useState<ElementObject | null>(null);
    const [isExpanded, setIsExpanded] = useState(false);
    const [isAllHighlighted, setIsAllHighlighted] = useState(false);
    const [textareaValues, setTextareaValues] = useState<string[]>(type.nodes.map(node => node.result.comment || ""));
    const [typeElements, setTypeElements] = useState<ElementObject[]>(type.nodes);
    const context = useContext(MyContext);
    if (context === null) {
      // handle the case where the context is null
      return null;
    }
    const { elementResults, setElementResults } = context;
    console.log(elementResults)
  
    const toggleCheck = () => {
        setIsAllHighlighted(!isAllHighlighted);
        console.log("Setter allhighlighted til " + isAllHighlighted);
        console.log("allhightlighted: " + isAllHighlighted);
        setCurrentHighlighted(null);
        highlightAll();
    };

    const updateJson = (elementObject: ElementObject, index: number) => {
        let newNodes = [...type.nodes];  // copy the array
        newNodes[index] = elementObject;  // replace the element
        setTypeElements(newNodes);  // update the state
        let elementResults : ElementResult[] = newNodes.map(node => node.result).flat();
        setElementResults(elementResults);
    };

    const storeText = (index: number) => {
        setTypeElements((prevTypeElements) => {
            const newNodes = [...prevTypeElements]; // copy the array
            if (newNodes[index]) {
                newNodes[index].result.comment = textareaValues[index]; // update the comment value
                updateJson(newNodes[index], index); // update the JSON with the updated element
            }
            return newNodes; // return the updated array
        });
    };
      
    const highlightAll = () => {
        messageSender.highlightAllWithType(type, isAllHighlighted);
    };
    return (
        <div className='collapsible-item'>
            <div className='collapsible-item-parent'>
                <div className={`item-header ${isExpanded ? 'pressed' : ''}`} onClick={() => setIsExpanded(!isExpanded)}>
                    <CollapsibleArrowButton isExpanded={isExpanded} />
                    <div className="buttons-text">
                        {type.name}
                    </div>
                    <ToggleButton isChecked={isAllHighlighted} onToggle={toggleCheck} text="Highlight All" />
                    <div className="total-buttons">
                        <p>{type.nodes.length}</p>
                    </div>
                </div>
            </div>

            {isExpanded && (
                <div className="collapsible-item-children">
                    {type.nodes.map((item, index) => (
                        <CollapsibleItemElement
                            type={type}
                            key={index}
                            thisElement={item}
                            highlightedElement={currentHighlighted}
                            setHighlightedElement={setCurrentHighlighted}
                            isAllHighlighted={isAllHighlighted}
                            setIsAllHighlighted={setIsAllHighlighted}
                            updateJson={updateJson}
                            index={index}
                        ><ElementAttributes
                                attributes={item.attributes}
                                title={item.title}
                                htmlString={item.htmlString}
                                selector={item.selector}
                                result={item.result} />
                            <SyntaxHighlighter language="html" style={vs}>
                                {item.htmlString}
                            </SyntaxHighlighter>
                            <div className="comment-box">

                        <textarea
                          className="textarea"
                          name="comment"
                          form="usrform"
                          value={textareaValues[index]}
                            onChange={(e) => setTextareaValues(prevValues => {
                                const newValues = [...prevValues];
                                newValues[index] = e.target.value;
                                return newValues;
                            })}
                        >
                            Enter text here...
                        </textarea>
                            </div>
                                <button className="store-text-button" onClick={() => storeText(index)}>
                                 Store Text
                                </button>
                        </CollapsibleItemElement>
                    ))}
                </div>
            )}
        </div>
    );
};

export const CollapsibleItemElement: React.FC<CollapsibleItemElementInterface> = ({
  type,
  thisElement,
  children,
  highlightedElement,
  isAllHighlighted,
  setHighlightedElement,
  setIsAllHighlighted,
  updateJson,
  index

}) => {

  const [isHighlighted, setIsHighlighted] = useState(false);
  const [isExpanded, setIsExpanded] = useState(false);


  useEffect(() => {
    setIsHighlighted((thisElement === highlightedElement) || isAllHighlighted);
  }, [highlightedElement, isAllHighlighted]);

  const handleCheckboxClick = (e: React.MouseEvent<HTMLInputElement>, update: string) => {
    e.stopPropagation(); // prevent event propagation
    if (update === "issue") {
        thisElement.result.issue = !thisElement.result.issue;
    } else if (update === "checked") {
        thisElement.result.checked = !thisElement.result.checked;
    }
    updateJson(thisElement, index);
};


  const toggleCheck = () => {
    //If we press the currently highlighted element, unhighlight it
    if (highlightedElement === thisElement) {
      setHighlightedElement(null);
      messageSender.highlightSingleMessage(thisElement, true);
      //} else if (isAllHighlighted && highlightedElement === null) {
    } else if (isAllHighlighted) {
      setIsAllHighlighted(false);
      messageSender.unhighlightAllAndHighlightSingleMessage(thisElement, type);
      //unhighlightAllAndHighligthSingle( );
      setHighlightedElement(thisElement);
    } else if (highlightedElement) {
      //Another element is highlighted, unhighlight it and highlight the new one
      messageSender.highlightAndRemovePreviousMessage(thisElement, highlightedElement);
      setHighlightedElement(thisElement);//Kan kanskje fjerne denne
    } else {
      //No element is highlighted, highlight the new one
      setHighlightedElement(thisElement);
      messageSender.highlightSingleMessage(thisElement, false);
    }
  };

  return (
    <div className="collapsible-item-child">
      <div className="collapsible-item">
        <div className={`item-header ${isExpanded ? 'pressed' : ''}`} onClick={() => setIsExpanded(!isExpanded)}>
          <div className="flex-item">
            <CollapsibleArrowButton isExpanded={isExpanded} />
            <div className="buttons-text">
              {thisElement.title}
            </div>
            <label htmlFor={`issueCheckbox-${index}`}>
                &nbsp;&nbsp;&nbsp;Error:
                  <input
                      id={`issueCheckbox-${index}`}
                      type="checkbox"
                      className="checkbox-custom"
                      tabIndex={2}
                      checked={thisElement.result.issue}
                      onClick={(e) => handleCheckboxClick(e, "issue")}
                  />
              </label>
              <label htmlFor={`viewedCheckbox-${index}`}>
                &nbsp;&nbsp;&nbsp;Checked:
                  <input
                      id={`viewedCheckbox-${index}`}
                      type="checkbox"
                      className="checkbox-custom"
                      tabIndex={2}
                      checked={thisElement.result.checked}
                      onClick={(e) => handleCheckboxClick(e, "checked")} />
              </label>
          </div>
          <ToggleButton isChecked={isHighlighted || isAllHighlighted} onToggle={toggleCheck} text="Jump to" />
        </div>
        <div className="content-data">
          {isExpanded && children}
        </div>
      </div>
    </div>
  );
};