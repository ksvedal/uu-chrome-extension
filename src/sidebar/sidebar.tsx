import React, {useContext, useEffect, useState} from 'react';
import { createRoot } from "react-dom/client";
import "../style/sidebar.css";
import { ElementObject, ElementResult, ElementType } from "./interfaces";
import { RegularButton } from './buttons';
import { MessageSender } from '../messageObjects/messageSender';
import { WebsiteScanner } from '../htmlParser/websiteScanner';
import ResultsHeader from './resultsHeader';
import { CollapsibleItemType } from './collapsibleItem';
import { MyContext } from './resultItemsContext';

export const Sidebar: React.FC = () => {
  const prefersDarkMode = window.matchMedia("(prefers-color-scheme: dark)").matches;
  const [darkMode, setDarkMode] = useState(prefersDarkMode);
  const [scanPage, setScanPage] = useState<ElementType[]>([]); // initialize scanPage state as an empty array
  const [websiteURL, setWebsiteURL] = useState<string>("");
  const [isAllHighlighted, setIsAllHighlighted] = useState<boolean>(false); // add this line
  const [currentHighlighted, setCurrentHighlighted] = useState<ElementObject | null>(null);
  const [elementResults, setElementResults] = useState<ElementResult[]>([]);
  const [index, setIndex] = useState<number[]>([]);
  const [thisElement, setThisElement] = useState<ElementObject | null>(null);
  const [testID, setTestID] = useState<string>("");

  const _message: MessageSender = new MessageSender();
  const _scan: WebsiteScanner = new WebsiteScanner();
  let dark: String = "light";

  useEffect(() => {
        const mediaQuery = window.matchMedia("(prefers-color-scheme: dark)");

        const handleChange = (event: MediaQueryListEvent) => {
            setDarkMode(event.matches);
        };

        mediaQuery.addEventListener("change", handleChange);

        return () => {
            mediaQuery.removeEventListener("change", handleChange);
        };
  }, []);

  const fetchData = () => {
    _message.scanPageMessage((response: ElementType[]) => {
      setScanPage(response); // update the state with the response data
      chrome.storage.local.set({ scanResults: response });
     
    });

    _scan.getWebsiteURL((url: string) => {
      setWebsiteURL(url); // update the state with the response data
    });
  };

  const toggleDarkMode = () => {
      setDarkMode(!darkMode);
    }

  return (
      <div className={`App ${darkMode ? "dark" : ""}`}>
      <div className='header-field'>
        <div className='extension-logo'>
          <img src="scan.png" alt="Extension Logo" />
        </div>
        <div className='extension-text'> <p>Button Seeker 2000</p></div>
        <button className={"dank-toggle-button float-right"} onClick={toggleDarkMode}> moon </button>
      </div>

        <div className={"row scan-page-field"}>
            <div className={"whitebox"}>
                <div className="col-8">
                    <div className='welcome-text'>
                        <p> Welcome to Button Seeker! Click the “Scan Page” to find all buttons</p>
                    </div>
                </div>
                <div className="col-4">
                    <RegularButton data-testid="scanPage" text="Scan Page" onClick={fetchData} />
                </div>
            </div>
        <div className="col-12">
            <MyContext.Provider value ={{elementResults, setElementResults}}>
                <div className={"whitebox"}>
                    <ResultsHeader url={websiteURL} isScanned={scanPage.length !== 0}/>
                </div>
            {/*for each element in ScanPage, creates a collapse menu with other nodes*/}
            {scanPage.map((item, index) =>
              <CollapsibleItemType key={index}
                type={item}
                setIsAllHighlighted={setIsAllHighlighted}
                setCurrentHighlighted={setCurrentHighlighted}
                isAllHighlighted={isAllHighlighted}
                parentIndex={index}
                thisElement={thisElement}
                url={websiteURL}
                testID={testID}
              >
              </CollapsibleItemType>)}
              </MyContext.Provider>
        </div>
        </div>

    </div>
  );
}

//export default Sidebar;
const root = createRoot(document.getElementById("root")!);

root.render(
  <React.StrictMode>
    <Sidebar />
  </React.StrictMode>
);