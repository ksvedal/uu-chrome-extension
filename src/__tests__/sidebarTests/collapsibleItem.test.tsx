// Import necessary dependencies and libraries
import React from "react";
import { render } from "@testing-library/react";
import { CollapsibleItemType } from "../../sidebar/collapsibleItem";
import '@testing-library/jest-dom/extend-expect';

// Describe the test suite for the 'CollapsibleItemType' component
describe("CollapsibleItemType", () => {
  // Define the test case: "should render without crashing"
  it("should render without crashing", () => {
    // Prepare the mock data for the 'type' prop
    const type = {
        name: "Test Type",
        nodes: [
          {
            title: "Test Element",
            htmlString: "",
            selector: "",
            result: {
              htmlString: "",
              correctText: "",
              name: "",
              comment: "",
              checked: false,
              url: "",
              testID: "",
              chromeVersion: null,
              chromeExtensionVersion: null,
              outcome: "",
            },
            attributes: [],
            isCommentVisible: false, // Add the isCommentVisible property here
          },
        ],
        selector: "some-selector", // Add a valid selector here
      };
      

    // Log a message indicating that the component is being rendered
    console.log("Rendering CollapsibleItemType...");

    // Render the 'CollapsibleItemType' component with the mock props
    const { container } = render(
      <CollapsibleItemType
        type={type}
        thisElement={null}
        isAllHighlighted={false}
        setIsAllHighlighted={() => {}}
        setCurrentHighlighted={() => {}}
        parentIndex={0}
        url=""
        testID=""
      />
    );

    // Log a message indicating that the component was rendered successfully
    console.log("CollapsibleItemType rendered successfully!");

    // Perform the test's primary assertion: Check if the container is in the document
    expect(container).toBeInTheDocument();

    // Optionally, you can use snapshot testing to compare the rendered output with a predefined snapshot
    expect(container).toMatchSnapshot();
  });
});





