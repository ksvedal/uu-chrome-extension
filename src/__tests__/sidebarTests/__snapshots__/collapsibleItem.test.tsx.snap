// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleItemType should render without crashing 1`] = `<div />`;
